internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.EnableVolumeReplicationResponse{}, nil
+}
+
+// DisableVolumeReplication disables replication on volume.
+func (cs *ControllerServer) DisableVolumeReplication(ctx context.Context,
+	req *replication.DisableVolumeReplicationRequest,
+) (*replication.DisableVolumeReplicationResponse, error) {
+	// get rbdVolume from VolumeID
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State == librbd.MirrorImageEnabled {
+		if !force && !mirroringInfo.Primary {
+			return nil, status.Error(codes.InvalidArgument, "image is in non-primary state")
+		}
+		err = rbdVol.disableImageMirroring(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to disable mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		// TODO check do we need to return abort after disabling the mirroring
+		return &replication.DisableVolumeReplicationResponse{}, nil
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageDisabled {
+		return nil, status.Errorf(codes.Aborted, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	return &replication.DisableVolumeReplicationResponse{}, nil
+}
+
+// PromoteVolume promotes the volume as primary.
+func (cs *ControllerServer) PromoteVolume(ctx context.Context,
+	req *replication.PromoteVolumeRequest,
+) (*replication.PromoteVolumeResponse, error) {
+	// get rbdVolume from VolumeID
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the force option
+	force, err := getForceOption(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
```

Comment:
This error message is not helpful. Always think about a user reading the error while troubleshooting an issue. Explain what is expected, and what the actual value is. Assist users with hints so that debugging a problem is easier.